### ISSUE TITLE
Inconsistent output dtype of arithmetic operators between single-element/multi-element tensors

### ISSUE THREAD

**Issue by xuhdev** [pytorch/pytorch#42473]
## 🐛 Bug

The output dtypes of arithmetic operators between a single-element tensor and a multi-element tensor, and that of arithmetic operators between two multi-element tensors, are inconsistent.

## To Reproduce

```
In [1]: import torch

In [2]: a=torch.tensor([190, 220], dtype=torch.uint8)

In [3]: b=torch.tensor([190, 210], dtype=torch.int8)

In [4]: a+b
Out[4]: tensor([124, 174], dtype=torch.int16)

In [5]: b=torch.tensor(210, dtype=torch.int8)

In [6]: a+b
Out[6]: tensor([144, 174], dtype=torch.uint8)
```

Note that the dtypes on lines 4 and 6 are different.

## Environment

PyTorch version: 1.7.0a0+ed44269
Is debug build: Yes
CUDA used to build PyTorch: Could not collect

OS: Debian GNU/Linux 10 (buster)
GCC version: (Debian 8.3.0-6) 8.3.0
CMake version: version 3.16.3

Python version: 3.7
Is CUDA available: No
CUDA runtime version: 10.1.243
GPU models and configuration: GPU 0: Quadro P400
Nvidia driver version: 440.100
cuDNN version: Could not collect

Versions of relevant libraries:
[pip3] numpy==1.18.1
[pip3] pytorch-sphinx-theme==0.0.24
[pip3] torch==1.7.0a0+ed44269
[conda] Could not collect

cc @nairbv

**Comment by nairbv** [comment]
This is intentional. We don't include the dtype of a scalar operand in type promotion if it's dtype is in the same "category" as the dimensioned operand. This rule allows, for example, `uint8_tensor * 5` to return a uint8 and not an int64.

See more examples here:
https://pytorch.org/docs/stable/tensor_attributes.html

**Comment by xuhdev** [comment]
@nairbv I understand when one operand is a scalar, but the examples above are on two tensors. I think the issue might be that `b` on Input 5 is seen as a scalar, not a tensor...

**Comment by nairbv** [comment]
> @nairbv I understand when one operand is a scalar, but the examples above are on two tensors. I think the issue might be that `b` on Input 5 is seen as a scalar, not a tensor...

We also differentiate zero-dim tensors from dimensioned tensors, similar to scalar literals:

https://pytorch.org/docs/stable/tensor_attributes.html
```
If the type of a scalar operand is of a higher category than tensor operands (where complex > floating > integral > boolean), we promote to a type with sufficient size to hold all scalar operands of that category.

If a zero-dimension tensor operand has a higher category than dimensioned operands, we promote to a type with sufficient size and category to hold all zero-dim tensor operands of that category.

If there are no higher-category zero-dim operands, we promote to a type with sufficient size and category to hold all dimensioned operands.
```
